cmd/relay/repomgr/repomgr.go

```diff
@@ -0,0 +1,398 @@
+			return nil, err
+		}
+
+		invTree := repoFragment.MST.Copy()
+		for _, op := range ops {
+			if err := atrepo.InvertOp(&invTree, &op); err != nil {
+				commitVerifyErrors.WithLabelValues(hostname, "inv").Inc()
+				return nil, err
+			}
+		}
+		computed, err := invTree.RootCID()
+		if err != nil {
+			commitVerifyErrors.WithLabelValues(hostname, "it").Inc()
+			return nil, err
+		}
+		if *computed != *c {
+			// this is self-inconsistent malformed data
+			commitVerifyErrors.WithLabelValues(hostname, "pd").Inc()
+			return nil, fmt.Errorf("inverted tree root didn't match prevData")
+		}
+		//logger.Debug("prevData matched", "prevData", c.String(), "computed", computed.String())
+
+		if prevRoot == nil {
+			commitVerifyOkish.WithLabelValues(hostname, "new").Inc()
+		} else if hasWarning {
+			commitVerifyOkish.WithLabelValues(hostname, "warn").Inc()
+		} else {
+			// TODO: would it be better to make everything "okish"?
+			// commitVerifyOkish.WithLabelValues(hostname, "ok").Inc()
+			commitVerifyOk.WithLabelValues(hostname).Inc()
+		}
+	} else {
+		// this source is still on old protocol without new prevData field
+		commitVerifyOkish.WithLabelValues(hostname, "old").Inc()
+	}
+
+	return repoFragment, nil
+}
+
+// TODO: lift back to indigo/atproto/repo util code?
+func ParseCommitOps(ops []*atproto.SyncSubscribeRepos_RepoOp) ([]atrepo.Operation, error) {
+	out := []atrepo.Operation{}
+	for _, rop := range ops {
+		switch rop.Action {
+		case "create":
+			if rop.Cid == nil || rop.Prev != nil {
+				return nil, fmt.Errorf("invalid repoOp: create")
+			}
+			op := atrepo.Operation{
+				Path:  rop.Path,
+				Prev:  nil,
+				Value: (*cid.Cid)(rop.Cid),
+			}
+			out = append(out, op)
+		case "delete":
+			if rop.Cid != nil || rop.Prev == nil {
+				return nil, fmt.Errorf("invalid repoOp: delete")
+			}
+			op := atrepo.Operation{
+				Path:  rop.Path,
+				Prev:  (*cid.Cid)(rop.Prev),
+				Value: nil,
+			}
+			out = append(out, op)
+		case "update":
+			if rop.Cid == nil || rop.Prev == nil {
+				return nil, fmt.Errorf("invalid repoOp: update")
+			}
+			op := atrepo.Operation{
+				Path:  rop.Path,
+				Prev:  (*cid.Cid)(rop.Prev),
+				Value: (*cid.Cid)(rop.Cid),
+			}
+			out = append(out, op)
+		default:
+			return nil, fmt.Errorf("invalid repoOp action: %s", rop.Action)
+		}
+	}
+	return out, nil
+}
+
+// VerifyCommitSignature get's repo's registered public key from Identity Directory, verifies Commit
+// hostname is just for metrics in case of error
+func (rm *RepoManager) VerifyCommitSignature(ctx context.Context, commit *atrepo.Commit, hostname string, hasWarning *bool) error {
+	if rm.directory == nil {
+		return nil
+	}
+	xdid, err := syntax.ParseDID(commit.DID)
+	if err != nil {
+		commitVerifyErrors.WithLabelValues(hostname, "sig1").Inc()
+		return fmt.Errorf("bad car DID, %w", err)
+	}
+	ident, err := rm.directory.LookupDID(ctx, xdid)
+	if err != nil {
+		if rm.AllowSignatureNotFound {
```

Comment:
should check if the `err` is an `ErrNotFound` before allowing here, right?

Reply:
maybe either `ErrDIDNotFound` (404) or `ErrDIDResolutionFailed` (5xx) should both be allow, but just now looking at the error return paths in the identity code there are a lot of different error conditions without a clear categorization (maybe the other broad category is "some part was there but invalid"). I think for now at this level "it didn't work, continue anyway" might be an okay strategy that we could refine later?

Reply:
I think if the PLC client can connect to the PLC directory and authoritatively determined that the PLC just doesn't exist at all, we should just drop the event, right? I don't think we should have a mechanism where PDS instances can spam the relay database with bogus DIDs which have never existed in any form.

It is maybe different if there are other identity errors, like network error connecting to the directory, in which case the DID might or might not exist, and we might make a different call to accept the DID even if we don't know it's disposition.